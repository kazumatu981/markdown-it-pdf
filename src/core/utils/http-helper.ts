import http from 'http';
import { type Logger } from '../../common';

//#region constants
const privatePortRange = {
    min: 49152,
    max: 65535,
};

const defaultListeningOptions = {
    retry: 10,
    range: privatePortRange,
};
//#endregion

//#region types and interfaces

/**
 * The server that is listening on a given port
 */
interface ListeningHttpServer {
    /**
     * The port that the server is listening on.
     */
    port: number;
    /**
     * The http server that is listening on the specified port.
     */
    httpServer: http.Server;
}

/**
 * The range of ports to search for
 */
export interface Range {
    /**
     * The minimum port to search for
     */
    min?: number;
    /**
     * The maximum port to search for
     */
    max?: number;
}

/**
 * Options for listening on a port
 */
export interface ListeningOptions {
    /**
     * The number of times to retry if the port is in use
     */
    retry?: number;
    /**
     * The range of ports to search for
     */
    range?: Range;
}

//#endregion

//#region functions

//#region exported functions
/**
 * Tries to listen on a given port. If the port is not specified, it tries to
 * listen on a random port within a specified range. If the specified port is
 * in use, it retries the specified number of times.
 *
 * @param {number | undefined} port The port to listen on. If not specified, it will try to listen on
 * a random port within the specified range.
 * @param {ListeningOptions | undefined} options The options for listening on a port.
 * @param {Logger | undefined} logger The logger to log the events.
 * @returns {Promise<ServerPort | undefined>} A promise that resolves to the port that the server is listening on
 * or undefined if the specified port is in use and cannot be listened on.
 */
export async function tryToListen(
    port?: number,
    options?: ListeningOptions,
    logger?: Logger
): Promise<ListeningHttpServer | undefined> {
    let listeningHttpServer = undefined;
    if (port != undefined) {
        logger?.debug('Try to Listen on port %d', port);
        listeningHttpServer = await tryToListenCore(port, logger);
    } else {
        logger?.debug('Try to Listen on port random mode');
        for (
            let i = 0;
            i < (options?.retry ?? defaultListeningOptions.retry);
            i++
        ) {
            port = getRandom(options?.range);
            listeningHttpServer = await tryToListenCore(port, logger);
            if (listeningHttpServer !== undefined) {
                break; // success
            }
        }
    }
    return listeningHttpServer;
}
//#endregion

//#region internal functions
/**
 * Generates a random number within a specified range.
 *
 * @param {Range | undefined} range - The range within which the random number
 * should be generated. If not specified, the function will use the default range
 * of 49152 to 65535.
 * @returns {number} A random number within the specified range.
 * @throws {Error} If the range is invalid (min > max).
 */
function getRandom(range?: Range): number {
    // Set the default range if not specified
    const candidate = {
        min: range?.min ?? defaultListeningOptions.range.min,
        max: range?.max ?? defaultListeningOptions.range.max,
    };

    // Check if the range is valid
    if (candidate.min > candidate.max) {
        throw new Error('Invalid range');
    }

    // Generate a random number within the range
    return Math.floor(
        Math.random() * (candidate.max - candidate.min) + candidate.min
    );
}

/**
 * Tries to listen on a given port. If the port is in use, it rejects the
 * promise.
 *
 * @param {number} port The port to listen on.
 * @param {Logger | undefined} logger The logger to log the events.
 * @returns {Promise<ServerPort | undefined>} A promise that resolves to the port that the server is listening on
 * or undefined if the specified port is in use and cannot be listened on.
 */
function tryToListenCore(
    port: number,
    logger?: Logger
): Promise<ListeningHttpServer | undefined> {
    // Create a new http server
    const server = http.createServer();

    // Create a promise that resolves to the port that the server is listening on
    // or undefined if the specified port is in use and cannot be listened on
    return new Promise((resolve) => {
        // Start listening on the specified port
        server.listen(port, () => {
            // Log the reserved port
            logger?.debug('reserved port: %d', port);
            // Resolve the promise with the port and the server
            resolve({
                port,
                httpServer: server,
            });
        });
        // If there is an error while listening on the port, log the error and
        // resolve the promise with undefined
        server.on('error', (error) => {
            logger?.debug('listen error: %s', error.message);
            logger?.trace(error.stack);
            logger?.warn('port %d may be used.', port);
            resolve(undefined);
        });
    });
}
//#endregion

//#endregion
