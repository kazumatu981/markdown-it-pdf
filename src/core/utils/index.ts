export * from './http-helper';
export * from './path-resolver';
