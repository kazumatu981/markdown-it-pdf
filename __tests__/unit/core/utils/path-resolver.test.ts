import { describe, it, expect } from '@jest/globals';
import {
    findFiles,
    filePathToUrl,
    normalizePath,
    buildTreeOfFiles,
} from '../../../../src/core/utils/path-resolver';
import { mockingTestDir, unmockingTestDir } from '../../../utils/test-dir';
import fsPromise from 'fs/promises';
import path from 'path';

describe('CoreLibrary Unit Tests - PathResolver', () => {
    it('Basic Test: Can find markdown files: recursive mode', async () => {
        mockingTestDir();
        const markdownFiles = await findFiles(
            'test',
            true,
            (filePath) => filePath.endsWith('.md'),
            (s) => s
        );
        unmockingTestDir();

        expect(markdownFiles).toContain(path.join('test', 'test.md'));
        expect(markdownFiles).toContain(path.join('test', 'sub', 'test.md'));
        expect(markdownFiles).toContain(path.join('test', 'sub2', 'test.md'));
        expect(markdownFiles).toContain(
            path.join('test', 'sub2', 'sub', 'test.md')
        );
        expect(markdownFiles).not.toContain(path.join('test', 'sample.txt'));
        expect(markdownFiles).toMatchSnapshot();
    });

    it('Basic Test: Can find markdown files: not recursive mode', async () => {
        mockingTestDir();
        const markdownFiles = await findFiles(
            'test',
            false,
            (filePath) => filePath.endsWith('.md'),
            (s) => s
        );
        unmockingTestDir();

        expect(markdownFiles).toContain(path.join('test', 'test.md'));
        expect(markdownFiles).not.toContain(
            path.join('test', 'sub', 'test.md')
        );
        expect(markdownFiles).not.toContain(
            path.join('test', 'sub2', 'test.md')
        );
        expect(markdownFiles).not.toContain(
            path.join('test', 'sub2', 'sub', 'test.md')
        );
        expect(markdownFiles).not.toContain(path.join('test', 'sample.txt'));
        expect(markdownFiles).toMatchSnapshot();
    });

    it('Error cases: Throws on root dir not found', async () => {
        mockingTestDir();
        expect(() =>
            findFiles(
                'test/does/not/exist',
                true,
                (filePath) => filePath.endsWith('.md'),
                (s) => s
            )
        ).rejects.toThrow();
        unmockingTestDir();
    });
    it('Error cases: Throws on file', async () => {
        mockingTestDir();
        expect(() =>
            findFiles(
                'test.md',
                true,
                (filePath) => filePath.endsWith('.md'),
                (s) => s
            )
        ).rejects.toThrow();
        unmockingTestDir();
    });
});

describe('CoreLibrary Unit Tests - filePathToUrl', () => {
    const rootDir = 'path';

    it('should convert a relative file path to a URL relative to the root directory', () => {
        const filePath = 'path/to/file.txt';
        const expectedUrl = '/to/file.txt';
        expect(filePathToUrl(rootDir, filePath)).toEqual(expectedUrl);
    });

    it('should throw an error if the file path is absolute', () => {
        const filePath = '/path2/to/file.txt';
        expect(() => filePathToUrl(rootDir, filePath)).toThrow(
            'Path is not in rootDir'
        );
    });

    it('should throw an error if the file path is not in the root directory', () => {
        const filePath = '../path2/to/file.txt';
        expect(() => filePathToUrl(rootDir, filePath)).toThrow(
            'Path is not in rootDir'
        );
    });
});

describe('CoreLibrary Unit Tests - normalizePath', () => {
    it('should normalize a relative file path', () => {
        const filePath = 'path/to/file.txt';
        const expectedNormalizedPath = 'path/to/file.txt';
        expect(normalizePath(filePath)).toEqual(expectedNormalizedPath);
    });

    it('should replace Windows-style path separators with POSIX-style separators', () => {
        const filePath = 'path\\to\\file.txt';
        const expectedNormalizedPath = 'path/to/file.txt';
        expect(normalizePath(filePath)).toEqual(expectedNormalizedPath);
    });

    it('should throw an error if the file path is absolute', () => {
        const filePath = '/path/to/file.txt';
        expect(() => normalizePath(filePath)).toThrow('Path must be relative');
    });
});

describe('CoreLibrary Unit Tests - buildTreeOfFiles', () => {
    it('should build a tree of directories', async () => {
        mockingTestDir();
        const files = ['path/to/file.txt', 'path/to/sub/file.txt'];
        await buildTreeOfFiles(files);
        // test that the directories were created
        const children = await fsPromise.readdir('path/to');
        expect(children).toContain('sub');

        unmockingTestDir();
    });
});
