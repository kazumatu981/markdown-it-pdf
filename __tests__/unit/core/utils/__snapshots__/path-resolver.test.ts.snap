// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreLibrary Unit Tests - findFiles Basic Test: Can find markdown files: not recursive mode 1`] = `
[
  "test\\test.md",
]
`;

exports[`CoreLibrary Unit Tests - findFiles Basic Test: Can find markdown files: recursive mode 1`] = `
[
  "test\\test.md",
  "test\\sub\\test.md",
  "test\\sub2\\test.md",
  "test\\sub2\\sub\\test.md",
]
`;
